### PR TITLE
http helper phase 2

### PR DESCRIPTION
**Pull request by Isabella-Pham**
## Description
Implementing phase 2 of the http helper function which sets the span error to 1 for custom status codes specified in the global config, stores query params, and stores request headers if specified in the integration config.

## Checklist
- [x] Create a status code parser to parse custom status codes that set the span error to 1
- [x] Add query_params and headers as parameters to the http helper function in trace_utils.py
- [x] Test updated http helper function

Update the http helper function to the following integrations:
- [x] aiohttp
- [x] asgi
- [x] bottle
- [x] django
- [x] falcon
- [x] flask
- [x] httplib
- [x] molten
- [x] pyramid
- [x] requests
- [x] sanic

Add query and/or header tests to the following integrations (as they do not have them):
- [x] bottle
- [x] molten
- [x] pyramid
- [x] sanic

### CODE REVIEW

File: ddtrace/contrib/trace_utils.py [hunk 1]
```diff
@@ -106,12 +107,46 @@ def ext_service(pin, config, default=None):
     return default
 
 
-def set_http_meta(span, config, method=None, url=None, status_code=None):
+def get_error_codes():
+    error_codes = []
+    try:
+        error_str = config.http_server.error_statuses
+    except AttributeError:
+        error_str = None
```
Comment (Kyle-Verhoog) [inline ddtrace/contrib/trace_utils.py:115]:
```suggestion
        return [[500, 599]]
```

File: ddtrace/contrib/trace_utils.py [hunk 2]
```diff
@@ -106,12 +107,46 @@ def ext_service(pin, config, default=None):
...
+    if error_str is None:
```
Comment (Kyle-Verhoog) [inline ddtrace/contrib/trace_utils.py:116]: This is not required with the above change

File: ddtrace/contrib/trace_utils.py [hunk 3]
```diff
@@ -106,12 +107,46 @@ def ext_service(pin, config, default=None):
     return default
 
 
-def set_http_meta(span, config, method=None, url=None, status_code=None):
+def get_error_codes():
```
Comment (Kyle-Verhoog) [inline ddtrace/contrib/trace_utils.py:110]: A docstring here would be helpful to get an idea of what is happening

File: ddtrace/contrib/trace_utils.py [hunk 4]
```diff
@@ -106,12 +107,46 @@ def ext_service(pin, config, default=None):
...
+    error_codes = []
+    try:
+        error_str = config.http_server.error_statuses
```
Comment (Kyle-Verhoog) [inline ddtrace/contrib/trace_utils.py:113]: Is this new functionality? If so I think we should move it to another PR.

File: ddtrace/contrib/trace_utils.py (hunk 3 above)
Comment (Kyle-Verhoog) [reply ddtrace/contrib/trace_utils.py:110]: Also a cleaner approach might be to have a function `is_error_code()` which simply returns whether a code is erroneous or not, rather than having to construct a list of all error ranges which then has to be checked.

File: ddtrace/contrib/trace_utils.py [hunk 5]
```diff
@@ -106,12 +107,46 @@ def ext_service(pin, config, default=None):
...
+            max_code = min_code
+        if min_code > max_code:
+            tmp = min_code
+            min_code = max_code
+            max_code = tmp
+        error_codes.append([min_code, max_code])
+    return error_codes
+
+
+def set_http_meta(span, integration_config, method=None, url=None, status_code=None, query=None, headers=None):
```
Comment (Kyle-Verhoog) [inline ddtrace/contrib/trace_utils.py:134]: `headers` should probably be `request_headers` since we'll want to support both `request_headers` and `response_headers`

File: ddtrace/contrib/requests/connection.py [hunk 6]
```diff
@@ -97,13 +97,11 @@ def _wrap_send(func, instance, args, kwargs):
             # Storing response headers in the span. Note that response.headers is not a dict, but an iterable
             # requests custom structure, that we convert to a dict
             if hasattr(response, "headers"):
-                store_response_headers(dict(response.headers), span, config.requests)
+                trace_utils.set_http_meta(span, config.requests, headers=dict(response.headers))
```
Comment (Kyle-Verhoog) [inline ddtrace/contrib/requests/connection.py:100]: These are response headers and not request headers so we probably don't want to set them here.

File: ddtrace/contrib/trace_utils.py (hunk 5 above)
Comment (Kyle-Verhoog) [reply ddtrace/contrib/trace_utils.py:134]: Should we also do `response_headers` in this PR? It might get a bit too big so it can wait for a "phase 3".

File: ddtrace/contrib/flask/middleware.py [hunk 7]
```diff
@@ -156,19 +155,31 @@ def _finish_span(self, span, exception=None):
...
         url = ""
+        query = ""
+        headers = ""
+
         if request:
-            store_request_headers(request.headers, span, config.flask)
             method = request.method
             endpoint = request.endpoint or code
             url = request.base_url or ""
+            query = request.query_string.decode()
```
Comment (Kyle-Verhoog) [inline ddtrace/contrib/flask/middleware.py:165]: We might want to safe-guard this `decode()`. Maybe something like

```suggestion
            query = request.query_string.decode() if hasattr(request.query_string, "decode") else request.query_string
```

File: ddtrace/contrib/trace_utils.py (hunk 4 above)
Comment (Isabella-Pham) [reply ddtrace/contrib/trace_utils.py:113]: Not new but not widely implemented. Here's some [documentation](https://github.com/DataDog/architecture/blob/master/rfcs/apm/integrations/status-error-mapping/rfc.md) on it.

File: ddtrace/contrib/trace_utils.py (hunk 5 above)
Comment (Isabella-Pham) [reply ddtrace/contrib/trace_utils.py:134]: I think adding response headers in this PR is good :D

File: ddtrace/contrib/flask/middleware.py [hunk 8]
```diff
@@ -125,6 +124,7 @@ def _process_response(self, response):
             return
 
         code = response.status_code if response else ""
+        trace_utils.set_http_meta(span, config.flask, status_code=code, response_headers=response.headers)
```
Comment (Kyle-Verhoog) [inline ddtrace/contrib/flask/middleware.py:127]: This looks to be new functionality which we don't have a test-case for. I'd be ok with pulling it out of this PR as well like you've done with the others 😄

File: ddtrace/contrib/requests/connection.py [hunk 9]
```diff
@@ -87,32 +86,36 @@ def _wrap_send(func, instance, args, kwargs):
...
                     response_headers = dict(getattr(response, "headers", {}))
-                    store_response_headers(response_headers, span, config.requests)
                 trace_utils.set_http_meta(
-                    span, config.requests, method=request.method.upper(), url=sanitized_url, status_code=status
+                    span,
+                    config.requests,
+                    method=request.method.upper(),
+                    url=sanitized_url,
+                    status_code=status,
+                    query=parsed_uri.query,
```
Comment (Kyle-Verhoog) [inline ddtrace/contrib/requests/connection.py:118]: Seems like we're missing the response_headers here? A test should probably have caught this 🤔

File: ddtrace/contrib/trace_utils.py [hunk 10]
```diff
@@ -106,12 +107,32 @@ def ext_service(pin, config, default=None):
...
+        span.meta[http.METHOD] = str(method)
+
     if url is not None:
-        span._set_str_tag(http.URL, url)
+        span.meta[http.URL] = stringify(url)
+
     if status_code is not None:
-        span._set_str_tag(http.STATUS_CODE, status_code)
-        if 500 <= int(status_code) < 600:
+        span.meta[http.STATUS_CODE] = str(status_code)
```
Comment (Kyle-Verhoog) [inline ddtrace/contrib/trace_utils.py:127]: Any particular reason for this change?

**Review by Kyle-Verhoog — COMMENTED** [review]
Almost there! Just caught a couple of things 🙂

File: ddtrace/contrib/requests/connection.py (hunk 9 above)
Comment (Isabella-Pham) [reply ddtrace/contrib/requests/connection.py:118]: They're stored in lines 95-101 :) using set_http_meta(), we have to store them in a different place in the code

File: ddtrace/contrib/trace_utils.py (hunk 10 above)
Comment (Isabella-Pham) [reply ddtrace/contrib/trace_utils.py]: nope that was done by mistake

File: ddtrace/contrib/requests/connection.py (hunk 9 above)
Comment (Kyle-Verhoog) [reply ddtrace/contrib/requests/connection.py:118]: ah missed it!

**Review by Kyle-Verhoog — APPROVED** [review]
👏
